src/QsCompiler/Compiler/ExternalRewriteSteps.cs

```diff
@@ -94,11 +132,18 @@ Uri WithFullPath(string file)
             { onDiagnostic?.Invoke(Errors.LoadError(ErrorCode.UnknownCompilerPlugin, new[] { file.LocalPath }, file.LocalPath)); }
 
             var rewriteSteps = ImmutableArray.CreateBuilder<LoadedStep>();
-            foreach (var (target, rewriteStepOptions) in foundDlls)
+            foreach (var (target, outputFolder) in foundDlls)
             {
                 var relevantTypes = new List<Type>();
                 Diagnostic LoadError(ErrorCode code, params string[] args) => Errors.LoadError(code, args, ProjectManager.MessageSource(target));
-                try { relevantTypes.AddRange(Assembly.LoadFrom(target.LocalPath).GetTypes().Where(t => t.IsAssignableFrom(typeof(IRewriteStep)))); }
+                Diagnostic LoadWarning(WarningCode code, params string[] args) => Warnings.LoadWarning(code, args, ProjectManager.MessageSource(target));
+                try
+                {
+                    var asmType = Assembly.LoadFrom(target.LocalPath).GetTypes();
+                    var exactInterfaceMatch = asmType.Where(t => typeof(IRewriteStep).IsAssignableFrom(t)); // inherited interface is defined in this exact dll
+                    var compatibleInterface = asmType.Where(t => t.GetInterfaces().Any(t => t.FullName == typeof(IRewriteStep).FullName)); // inherited interface may be defined in older compiler version
+                    relevantTypes.AddRange(exactInterfaceMatch.Concat(compatibleInterface)); 
```

Comment:
This double Where + Concat combo could cause duplicate entries in the resulting list. Do you want entries that satisfy both conditions to appear twice? Even if the conditions are mutually exclusive, if you don't want the duplicate behavior it is best to avoid this pattern either by having one Where with both conditions, or two consecutive Where's on the same enumerable.